applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.cpp

```diff
@@ -13,108 +13,85 @@
+    Element::IntegrationMethod AdjointBasePotentialFlowElement<TPrimalElement>::GetIntegrationMethod() const
     {
-        KRATOS_TRY
-          return Kratos::make_intrusive<AdjointPotentialFlowElement<TPrimalElement>>(NewId, GetGeometry().Create(ThisNodes), pProperties);
-        KRATOS_CATCH("");
-    }
-
-    template <class TPrimalElement>
-    Element::Pointer AdjointPotentialFlowElement<TPrimalElement>::Create(IndexType NewId, GeometryType::Pointer pGeom, PropertiesType::Pointer pProperties) const 
-    {
-        KRATOS_TRY
-            return Kratos::make_intrusive<AdjointPotentialFlowElement<TPrimalElement>>(NewId, pGeom, pProperties);
-        KRATOS_CATCH("");
-    }
-
-    template <class TPrimalElement>
-    Element::Pointer AdjointPotentialFlowElement<TPrimalElement>::Clone(IndexType NewId, NodesArrayType const& ThisNodes) const 
-    {
-        KRATOS_TRY
-        return Element::Pointer(new AdjointPotentialFlowElement(NewId, GetGeometry().Create(ThisNodes), pGetProperties()));
-        KRATOS_CATCH("");
+        return mpPrimalElement->GetIntegrationMethod();
     }
 
     template <class TPrimalElement>
-    Element::IntegrationMethod AdjointPotentialFlowElement<TPrimalElement>::GetIntegrationMethod() const 
+    void AdjointBasePotentialFlowElement<TPrimalElement>::Initialize()
     {
-        return mpPrimalElement->GetIntegrationMethod();
-    }
-  
-    template <class TPrimalElement>
-    void AdjointPotentialFlowElement<TPrimalElement>::Initialize() 
-    {   
         mpPrimalElement->Initialize();
     }
 
     template <class TPrimalElement>
-    void AdjointPotentialFlowElement<TPrimalElement>::InitializeSolutionStep(ProcessInfo& rCurrentProcessInfo) 
+    void AdjointBasePotentialFlowElement<TPrimalElement>::InitializeSolutionStep(ProcessInfo& rCurrentProcessInfo)
     {
         mpPrimalElement->Data() = this->Data();
         mpPrimalElement->Set(Flags(*this));
         mpPrimalElement->InitializeSolutionStep(rCurrentProcessInfo);
     }
 
     template <class TPrimalElement>
-    void AdjointPotentialFlowElement<TPrimalElement>::FinalizeSolutionStep(ProcessInfo& rCurrentProcessInfo) 
+    void AdjointBasePotentialFlowElement<TPrimalElement>::FinalizeSolutionStep(ProcessInfo& rCurrentProcessInfo)
     {
 
     }
 
     template <class TPrimalElement>
-    void AdjointPotentialFlowElement<TPrimalElement>::CalculateLocalSystem(MatrixType& rLeftHandSideMatrix,
+    void AdjointBasePotentialFlowElement<TPrimalElement>::CalculateLocalSystem(MatrixType& rLeftHandSideMatrix,
                                       VectorType& rRightHandSideVector,
-                                      ProcessInfo& rCurrentProcessInfo) 
+                                      ProcessInfo& rCurrentProcessInfo)
     {
         mpPrimalElement->CalculateLocalSystem(rLeftHandSideMatrix,
```

Comment:
I guess that `mpPrimalElement` is a pointer to the base element implementing the formulation (incompressible / compressible potential flow elements). If so it is, do you need a pointer to it? Isn't it sufficient to do `TPrimalElement::CalculateLocalSystem()`?

Reply:
So this follows @armingeiser and @msandre design if im not mistaken, please correct me if im wrong:

TPrimalElement is not constructed (with an index, a geometry, etc) so we can not call directly the methods from it, as it has no data.

Here is where the adjoint element is defined:

https://github.com/KratosMultiphysics/Kratos/blob/1df0f52127ddbaf9cdbbef62c368fe4bf9540fcd/applications/CompressiblePotentialFlowApplication/compressible_potential_flow_application.h#L104-L105

At the constructor of the adjoint element, we construct the primal element and  we store a pointer (once initiazlied with an index, geometry, etc..):

https://github.com/KratosMultiphysics/Kratos/blob/1df0f52127ddbaf9cdbbef62c368fe4bf9540fcd/applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.h#L53-L58

So to answer your question, I would say that it is not possible to use `TPrimalElement::CalculateLocalSystem()` since TPrimalElement hasn't been constructed yet.

Reply:
Right, `TPrimalElement` is just the class, but the function needs to be called at the instance.

Reply:
Thanks for the explanation.

Now I see. I though you followed a similar design to the ones in the FluidDynamics, in which we derive from a base element that implements the formulation instead of deriving from the base `Element` in the core.

Just for me to know, is there any reason to not derive from the compressible / incompressible elements? Even though the elements are not shared pointers anymore, it may be still worth for the sake of memory consumption.

Reply:
The main reason is that I followed the StructuralApp design to implement the adjoint in the potential flow, especially to use their StructuralAdjoint utilitities (sensitivity builder) and processes (replace multiple element and conditions). Now we are using both processes from the core so a different scheme could be considered. Also the ShapeOpt examples were running with the structural case so it was more straight forward to follow the structural approach.

Deriving from the compressible/incomressible elements (and the embedded ones) would require to have a complete adjoint for each of those if Im not mistaken (with repeated code). That would be (in the short term) 4 primal elements + 4 adjoint elements, which is not too much actually. It wouldn't be too much work either since the adjoint elements do not have many functions.

The advantage of the current approach is that a new element formulation would work out of the box with the same adjoint element without having to implement anything else. For instance, once we added the compressible element, we didn't have to add new functions to the adjoint element since it could use the same one that was currently implemented. We only needed to define it in the application with its correct template.

In short I think that the current implementation is better and cleaner if more elements/formulations are planned to be added. Deriving from the primals direclty could be considered however to improve performance if it is worth to do so.

Not sure if there is another big reason to use template `TPrimalElement` instead of deriving from the primals @armingeiser ?

Reply:
What I mean is that by doing
```c++
template <class TPrimalElement>
class AdjointElement : public TPrimalElement {...}
```
instead of 
```c++
template <class TPrimalElement>
class AdjointElement : public Element {...}
```
you could save the overhead of one pointer per element with no behavior change. You will only need to forward declare for each one of the primal elements, as you are doing right now.

I was just wondering. There should be a reason because of which you selected the current design.

Reply:
I haven't followed the discussion closely but I remember discussion this many times with @armingeiser 
afair the first version of the adjoint elements did derive from the primal version, but there were some issues that could not really be solved (@armingeiser maybe you remember what it was?)
Hence the decision was taken to change to the current approach, which to my knowledge works best so far
also ping @msandre since he was also involved in this design

Reply:
Yes. In our case we can not derive from the primal element, because we need the primal to work as is - It`s a bit complicated to explain, but i try:
The main reason is that the adjoint sensitivity analysis is composed of the adjoint problem (adjoint elements) itself and the derivatives of the system (primal elements). In the current design all this is done using the same model part, that only has the adjoint elements. This gets complicated by the fact that we are calculating most of the derivatives with finite differences, e.g. we need to be able to calculate stresses in the primal element using the primal variables as dofs, while the adjoint elements have the adjoint variables as dofs. If an adjoint element derives from the primal element and overwrites the GetDofs, GetValues methods, in the example given, the stresses would be calculated using the `ADJOINT_DISPLACEMENT`.

Reply:
Thanks for the clear explanation @armingeiser . I think that more or less I got the idea. Honestly, no better design comes to my mind neither.